src/Voron/Impl/LowLevelTransaction.cs

```diff
@@ -1290,22 +1027,27 @@ private void CommitStage1_CompleteTransaction()
                 FreePage(_pagesToFreeOnCommit.Pop());
             }
 
-            if (_state.NextPageNumber <= 1)
+            if (GetNextPageNumber() <= 1)
                 ThrowNextPageNumberCannotBeSmallerOrEqualThanOne();
 
-            _txHeader.LastPageNumber = _state.NextPageNumber - 1;
+            _txHeader.LastPageNumber = GetNextPageNumber() - 1;
+            
             ref var rootHeader = ref _txHeader.Root;
-            _state.Root.CopyTo(ref rootHeader);
+            _envRecord.Root.CopyTo(ref rootHeader);
 
             _txHeader.TxMarker |= TransactionMarker.Commit;
 
             LastChanceToReadFromWriteTransactionBeforeCommit?.Invoke(this);
+
+            _env.Journal.Applicator.OnTransactionCommitted(this);
```

Comment:
This was called during `CommitStage3_` previously, while it's much earlier now. In particular it's called _before_ the actual write to the journal now.

Reply:
Also this call will executed `scratch.File.Free(tx, page)` under the covers (via `UpdateJournalStateUnderWriteTransactionLock()`). We use current tx there but it's not already committed yet. We need to double check that.

Reply:
BTW was there any issue with calling `_env.Journal.Applicator.OnTransactionCommitted(this);` during `CommitStage3_`?

Reply:
Yes, that is intentional move.
Note that at this point, this is not related to the _transactional_ state.
And rolling back the transaction _will not_ affect the changes it makes.

It basically updates the journals and scratches you have - and that is not rolled back.